resources/js/legacy-irc-key/controller.ts

```diff
@@ -0,0 +1,71 @@
+// Copyright (c) ppy Pty Ltd <contact@ppy.sh>. Licensed under the GNU Affero General Public License v3.0.
+// See the LICENCE file in the repository root for full licence text.
+
+import LegacyIrcKeyJson from 'interfaces/legacy-irc-key-json';
+import { route } from 'laroute';
+import { action, makeObservable, observable, reaction, runInAction } from 'mobx';
+import { onError } from 'utils/ajax';
+
+interface State {
+  legacy_irc_key: LegacyIrcKeyJson | null;
+}
+
+export default class Controller {
+  @observable state;
+  private stateSyncDisposer;
+  private xhrCreate?: JQuery.jqXHR<LegacyIrcKeyJson>;
+  private xhrDelete?: JQuery.jqXHR<void>;
+
+  get isCreating() {
+    return this.xhrCreate != null;
+  }
+
+  get isDeleting() {
+    return this.xhrDelete != null;
+  }
+
+  get key() {
+    return this.state.legacy_irc_key;
+  }
+
+  constructor(private container: HTMLElement) {
+    this.state = JSON.parse(container.dataset.state ?? '') as State;
+
+    makeObservable(this);
+
+    this.stateSyncDisposer = reaction(
+      () => JSON.stringify(this.state),
+      (stateString) => this.container.dataset.state = stateString,
+    );
+  }
+
+  createKey() {
```

Comment:
also needs `@action` to go with the `@observable` to properly show the spinner or disable button, same for the other create/delete keys